### PR TITLE
Add visibility checkbox description

### LINKED ISSUE #29819: Visibility (catalog, everywhere, …): there should be a dynamic help text above the radio buttons
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing features request](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3AImprovement%2CFeature%2CDeveloper+Feature+) and found no previous suggestion of this feature.

### Is your feature request related to a problem?

Adapt the content help text of the visibility option in the product page depending on option selected by the user

### Describe the solution you'd like
The help text is below the option.
- Everywhere:  Help text: Customers can access the product by browsing the catalog, using the search bar, or the link.
- Catalog only: Help text: Customers can access the product only by browsing the catalog. This is particularly useful to avoid displaying too many similar products in a search.
- Search only: Help text: Customers can access the product only by using the search bar.
- Nowhere: Help text: Only customers with the link can access the product

#### Acceptance test 

//To do 

### Do you plan to work on this feature?

- [X] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [x] I'm willing to provide any wireframes or design assets required for this feature.
- [ ] I'm willing to submit a Pull Request that implements this feature.
- [X] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

### PR DESCRIPTION
**Pull request by FabienPapet**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | This PR add an dynamic help on the attribute that changes regarding the value you have selected.
| Type?             | new feature
| Category?         | BO
| BC breaks?        | yes (on experimental)
| Deprecations?     | no
| Fixed ticket?     | Fixes #29819

### CODE REVIEW

File: src/Core/Domain/Product/ValueObject/ProductVisibility/AbstractVisibility.php
Comment (matks) [inline src/Core/Domain/Product/ValueObject/ProductVisibility/AbstractVisibility.php]: Interesting idea 👍

**Review by matks — COMMENTED** [review]
PHPStan is not happy but I think the Visibility classes and interface you introduce are a good idea

File: admin-dev/themes/new-theme/js/pages/product/edit/product-options-manager.ts
Comment (NeOMakinG) [inline admin-dev/themes/new-theme/js/pages/product/edit/product-options-manager.ts]: Using a generic like `document.querySelector<HTMLDivElement>('.js-description')` would be safer and cleaner, and check if it exist right after

Sometime in our paradigm, we're tempted to cast the type as anyway the HTML is not generated by any JS lib meaning that we've no control over it, but when we can because the code part is pretty small, let's use the cleaner method!

File: src/PrestaShopBundle/Form/Admin/Sell/Product/Options/VisibilityType.php
Comment (kpodemski) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/Options/VisibilityType.php]: Worth checking with @jolelievre if this is something that catalog exporter can handle

Comment (FabienPapet) [reply src/PrestaShopBundle/Form/Admin/Sell/Product/Options/VisibilityType.php]: fixed

**Review by yanmakouf — DISMISSED** [review]
Seems good to me :)

File: admin-dev/themes/new-theme/js/pages/product/edit/product-options-manager.ts
Comment (jolelievre) [inline admin-dev/themes/new-theme/js/pages/product/edit/product-options-manager.ts]:
```suggestion
    const defaultSelectedInput = document.querySelector<HTMLInputElement>(`${ProductMap.options.visibilityRadio}:checked`);
```

Comment (jolelievre) [reply admin-dev/themes/new-theme/js/pages/product/edit/product-options-manager.ts]: We prefer the use of the generic (like you did with `descriptionField`), because it doesn't fake/force the type. The only drawback is that it can return null (when the element is not found) but it actually means that the returned type is closer to the real use case

File: src/Core/Form/ChoiceProvider/ProductVisibilityChoiceProvider.php
Comment (jolelievre) [inline src/Core/Form/ChoiceProvider/ProductVisibilityChoiceProvider.php:35]: You shouldn't remove this service, most of our form types rely on this interface (or at least use them for injection). 
In this case to have some additional parameters (your data attribute), you should use `FormChoiceAttributeProviderInterface` in this same service and you return for each choice an array of attributes in your case something like:

```php
return [
           $this->trans('Everywhere', 'Admin.Catalog.Feature') => [
                'data-description' => $this->trans('Customers can access the product by browsing the catalog, using the search bar, or the link.', 'Admin.Catalog.Feature'),
            ]
            ....
        ];
```

You must use the translated label as the key of your array, because that's what the ChoiceType field expects.

File: src/PrestaShopBundle/Form/Admin/Sell/Product/Options/VisibilityType.php
Comment (jolelievre) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/Options/VisibilityType.php:54]: One improvement you could do here though, is replace this injection of `FormChoiceProviderInterface` with:
```php
    array $visibiltyChoices,
    array $visibiltyAttributeChoices
```
So that in your service definition you can use `@=service('prestashop.core.form.choice_provider.product_visibility_choice_provider').getChoices()` and `@=service('prestashop.core.form.choice_provider.product_visibility_choice_provider').getChoicesAttributes()`

File: src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/prestashop_ui_kit_base.html.twig
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/prestashop_ui_kit_base.html.twig]: Nice! 👍

File: admin-dev/themes/new-theme/js/pages/product/edit/product-options-manager.ts
Comment (jolelievre) [inline admin-dev/themes/new-theme/js/pages/product/edit/product-options-manager.ts]:
```suggestion
    const inputs = document.querySelectorAll<HTMLInputElement>(ProductMap.options.visibilityRadio) ?? [];
```
You also have a generic on `querySelectorAll` this way you don't need to explicitly type `inputs` because TypeScript can infer the type accurately

The ` ?? []` at the end prevents you from checking that `inputs` is not null

File: admin-dev/themes/new-theme/js/pages/product/edit/product-options-manager.ts
Comment (jolelievre) [inline admin-dev/themes/new-theme/js/pages/product/edit/product-options-manager.ts]:
```suggestion
    inputs.forEach((input: HTMLInputElement) => {
```
This is optional, since TypeScript is supposed to infer the type correctly

**Review by jolelievre — COMMENTED** [review]
Thanks @FabienPapet 

But you shouldn't remove the choice provider service, it's part of our convention for form types. It reduces the form type code size, it makes the choice providers reusable, they can even be unit tested (even if they are not all tested), and it allows decoration/extension/replacement of these services by modules, which is easier than customizing the form type itself.

**Review by atomiix — DISMISSED** [review]

File: admin-dev/themes/new-theme/js/pages/product/edit/product-options-manager.ts
Comment (jolelievre) [inline admin-dev/themes/new-theme/js/pages/product/edit/product-options-manager.ts]: `'.js-description'` should be placed into the product component mapping

File: src/PrestaShopBundle/Form/Admin/Sell/Product/Options/VisibilityType.php
Comment (jolelievre) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/Options/VisibilityType.php]:
```suggestion
                'help_attr' => ['class' => 'js-visibility-description'],
```

Comment (jolelievre) [reply src/PrestaShopBundle/Form/Admin/Sell/Product/Options/VisibilityType.php]: I suggest a more specific class here, `description` is a bit too generic and could be common with other fields

**Review by jolelievre — COMMENTED** [review]
Just a few adjustments but all the rest is ok

File: admin-dev/themes/new-theme/js/pages/product/edit/product-options-manager.ts
Comment (jolelievre) [reply admin-dev/themes/new-theme/js/pages/product/edit/product-options-manager.ts]: @FabienPapet only this selector class left to improve and the PR can go to QA

**Review by jolelievre — CHANGES_REQUESTED** [review]

**Review by jolelievre — DISMISSED** [review]
Thanks @FabienPapet

**Review by nicosomb — DISMISSED** [review]

**Review by florine2623 — COMMENTED** [review]
Hello @FabienPapet ,

Automated tests are green ✅ 

Checked on PPV2, PR is well installed, I have rebuilt the assets and cleared cache, but I don't see any changes:

https://user-images.githubusercontent.com/16019289/204764672-3da3261e-95c8-4230-83ae-96a0c6af92ff.mov

Am I missing something ?

**Comment by FabienPapet** [comment]
@florine2623 I fixed the problem, can you check again ? 

@PrestaShop/committers needs approvals again please 🙏

**Comment by hibatallahAouadni** [comment]
This PR needs Committers approval before putting the `Waiting for QA` label 😉

**Review by nicosomb — DISMISSED** [review]

**Review by paulnoelcholot — COMMENTED** [review]
Hello @FabienPapet,

I have tester this PR and I found a bug : 
image
As seen in the issus : https://github.com/PrestaShop/PrestaShop/issues/29819

The wording is not in the right place, it should be on top. 

Also the CI is down.

Why is the text enclosed in a box rather than on a line? This will avoid having the rest of the page impacted if the text is too long (currently 1 to 3 lines)

Thanks you !

**Review by jolelievre — COMMENTED** [review]
Hi @FabienPapet 

I think you need to rebase your PR a bug has been fixed on develop which failed all the CIs
And there's a feedback from @paulnoelcholot about the position of the description that should be fixed as well

**Review by zuk3975 — APPROVED** [review]

**Review by lartist — APPROVED** [review]

**Review by jolelievre — APPROVED** [review]
Thanks @FabienPapet

**Comment by AureRita** [comment]
Hi @FabienPapet 

Thank for your PR, I have the same bug as Paul-Noel, as you can see : 
image

even if I do a `make assets` , I didn't have the text on the top or in one line 

Thank you

**Comment by atomiix** [comment]
@AureRita @paulnoelcholot I don't see where in the issue is mentioned that the help text should be on top? It even surprises me as every help text of every form element is usually under the field 🤔

**Comment by paulnoelcholot** [comment]
Previously, the message was at the top. To keep the same logic, it would therefore be necessary to keep this behavior. However, your remark is also relevant and it will be interesting to have a UX and/or PM opinion on the matter.
@MatShir ?

**Comment by MatShir** [comment]
Hi guy, 
I forgot to update the issue, after discussion we decide to put it at the bottom. 
@paulnoelcholot. Have you found other issue on this PR?

**Comment by paulnoelcholot** [comment]
It's good for me !

**Review by paulnoelcholot — APPROVED** [review]

**Comment by kpodemski** [comment]
Thanks @FabienPapet

**Pull request merged**
